scheduler/src/cook/mesos/data_locality.clj

```diff
@@ -1,48 +1,200 @@
 (ns cook.mesos.data-locality
-  (:require [cook.config :as config]
+  (:require [cheshire.core :as cheshire]
+            [clj-http.client :as http]
+            [clj-time.coerce :as tc]
+            [clj-time.core :as t]
+            [clj-time.format :as tf]
+            [clojure.set :as set]
+            [clojure.tools.logging :as log]
+            [cook.config :as config]
+            [cook.mesos.util :as util]
+            [datomic.api :as d]
+            [metrics.histograms :as histograms]
+            [metrics.timers :as timers]
             [plumbing.core :as pc])
-  (:import com.netflix.fenzo.VMTaskFitnessCalculator
-           com.netflix.fenzo.plugins.BinPackingFitnessCalculators))
+  (:import com.google.common.cache.Cache
+           com.netflix.fenzo.VMTaskFitnessCalculator
+           com.netflix.fenzo.plugins.BinPackingFitnessCalculators
+           java.util.UUID))
+
+(def partition-date-format (:basic-date tf/formatters))
+
+(defonce ^Cache job-uuid->dataset-maps-cache (util/new-cache))
+
+(defn- make-partition-map
+  [partition-type partition]
+  (let [format-date (fn format-date [date] (tf/unparse partition-date-format (tc/from-date date)))]
+    (case partition-type
+      "date" {"begin" (format-date (:dataset.partition/begin partition))
+              "end" (format-date (:dataset.partition/end partition))})))
+
+(defn- make-dataset-maps
+  [job]
+  (->> job
+       :job/datasets
+       (map (fn [{:keys [dataset/partitions dataset/partition-type dataset/parameters]}]
+              (let [partitions (when (not (empty? partitions))
+                                 (->> partitions
+                                      (map (partial make-partition-map partition-type))
+                                      (into #{})))]
+                (cond-> {:dataset (into {} (map (fn [p] [(:dataset.parameter/key p) (:dataset.parameter/value p)])
+                                                parameters))}
+                  partitions (assoc :partitions partitions)))))
+       (into #{})))
+
+(defn get-dataset-maps
+  "Returns the (possibly cached) datasets for the given job"
+  [job]
+  (util/lookup-cache! job-uuid->dataset-maps-cache
+                      :job/uuid
+                      make-dataset-maps
+                      job))
+
+(histograms/defhistogram [cook-mesos data-locality cost-staleness])
+
+(let [datasets->host-name->cost-atom (atom {})
+      datasets->last-update-time-atom (atom {})]
```

Comment:
Should these be `dataset->...` (singular)?

Reply:
It's a set-of-datasets